leader-election-impl/src/main/java/com/palantir/paxos/LeaderPingResult.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.paxos;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.derive4j.Data;
+
+import com.palantir.leader.PaxosLeaderElectionEventRecorder;
+
+@Data
+public abstract class LeaderPingResult {
+
+    interface Cases<R> {
+        R pingTimedOut();
+        R pingReturnedTrue();
+        R pingReturnedFalse();
+        R pingCallFailure(Throwable exception);
+    }
+
+    public abstract <R> R match(Cases<R> cases);
+
+    public static LeaderPingResult fromBooleanResult(boolean result) {
+        if (result) {
+            return LeaderPingResults.pingReturnedTrue();
+        } else {
+            return LeaderPingResults.pingReturnedFalse();
+        }
+    }
+
+    public void recordEvent(PaxosLeaderElectionEventRecorder eventRecorder) {
+        LeaderPingResults.caseOf(this)
+                .pingTimedOut(wrap(eventRecorder::recordLeaderPingTimeout))
+                .pingReturnedFalse(wrap(eventRecorder::recordLeaderPingReturnedFalse))
+                .pingCallFailure(wrap(eventRecorder::recordLeaderPingFailure))
+                .otherwise_(null);
+    }
+
+    public boolean isSuccessful() {
+        return LeaderPingResults.caseOf(this)
+                .pingReturnedTrue_(true)
+                .otherwise_(false);
```

Comment:
Possibly just a matter of taste, but I really don't like the trailing underscores. I'd prefer the `otherwise(() -> false);` that this actually is under the hood

Reply:
you get used to it pretty quickly, having to write suppliers/functions for constants gets pretty annoying/visual cruft. 

Agree underscores aren't ideal, but I think pointless suppliers are worse.